Sources/PackageGraph/RegistryPackageContainer.swift

```diff
@@ -0,0 +1,242 @@
+                            .lazy
+                            .filter { version in
+            self.validToolsVersionsCache.memoize(version) {
+                self.isToolsVersionCompatible(at: version)
+            }
+        })
+    }
+
+    public func versionsDescending() throws -> [Version] {
+        self.versionsDescendingCache
+    }
+
+    public func versionsAscending() throws -> [Version] {
+        self.versionsDescendingCache.reversed()
+    }
+
+    public func getDependencies(at version: Version, productFilter: ProductFilter) throws -> [PackageContainerConstraint] {
+        return try self.manifest(for: version).dependencyConstraints(productFilter: productFilter, mirrors: mirrors)
+    }
+
+    public func getDependencies(at revision: String, productFilter: ProductFilter) throws -> [PackageContainerConstraint] {
+        throw RegistryError.invalidOperation
+    }
+
+    public func getUnversionedDependencies(productFilter: ProductFilter) throws -> [PackageContainerConstraint] {
+        return []
+    }
+
+    public func getUpdatedIdentifier(at boundVersion: BoundVersion) throws -> PackageReference {
+        guard case .version(let version) = boundVersion else {
+            assertionFailure("Unexpected type requirement \(boundVersion)")
+            return identifier
+        }
+
+        return try identifier.with(newName: self.manifest(for: version).name)
+    }
+}
+
+// MARK: -
+
+public class RegistryPackageContainerProvider {
+    let manifestLoader: ManifestLoaderProtocol
+    let mirrors: DependencyMirrors
+    let currentToolsVersion: ToolsVersion
+    let toolsVersionLoader: ToolsVersionLoaderProtocol
+
+    private var containerCache = ThreadSafeKeyValueStore<PackageReference, PackageContainer>()
+
+    public init(
+        mirrors: DependencyMirrors = DependencyMirrors(),
+        manifestLoader: ManifestLoaderProtocol,
+        currentToolsVersion: ToolsVersion = ToolsVersion.currentToolsVersion,
+        toolsVersionLoader: ToolsVersionLoaderProtocol = ToolsVersionLoader()
+    ) {
+        self.mirrors = mirrors
+        self.manifestLoader = manifestLoader
+        self.currentToolsVersion = currentToolsVersion
+        self.toolsVersionLoader = toolsVersionLoader
+    }
+}
+
+extension RegistryPackageContainerProvider: PackageContainerProvider {
+    public func getContainer(
+        for identifier: PackageReference,
+        skipUpdate: Bool,
+        on queue: DispatchQueue,
+        completion: @escaping (Result<PackageContainer, Swift.Error>) -> Void
+    ) {
+        assert(identifier.kind == .remote)
+
+        if let container = containerCache[identifier] {
+            queue.async {
+                completion(.success(container))
+            }
+        } else {
+            RegistryManager.discover(for: identifier) { result in
+                switch result {
+                case .success(let registryManager):
+                    RegistryPackageContainer.create(
```

Comment:
I think we should do this entire operation on the queue (untested code):

```
queue.async {
  RegistryPackageContainer.create(
                         for: identifier,
                         mirrors: self.mirrors,
                         registryManager: registryManager,
                         manifestLoader: self.manifestLoader,
                         toolsVersionLoader: self.toolsVersionLoader,
                         currentToolsVersion: self.currentToolsVersion
                     ) { result in
                         switch result {
                         case .success(let container):
                             self.containerCache[identifier] = container
                             completion(.success(container))
                         case .failure(let error):
                             completion(.failure(error))
                         }
                     }
}

```

Reply:
alternatively, if RegistryPackageContainer.create is a sync, pass the queue to it and then we can simplify further